File: 5-Problem-A-Day/259. 3Sum Smaller.java
class Solution {
    int ans;

    public int threeSumSmaller(int[] nums, int target) {
        Arrays.sort(nums);
        int sum = 0;
        for (int i = 0; i < nums.length - 2; i++) {
            sum += twoSumSmaller(nums, i + 1, nums.length - 1, target - nums[i]);
        }
        return sum;
    }

    private int twoSumSmaller(int[] nums, int start, int end, int target) {
        int sum = 0;
        while (start < end) {
            if (nums[start] + nums[end] < target) {
                sum += end - start;
                start += 1;
            } else
                end -= 1;
        }
        return sum;
    }
}

File: 2021/259. 3Sum Smaller.cpp
class Solution {
public:
    int threeSumSmaller(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int total = 0;
        for (int i = 0; i < nums.size(); i++) {
            total += twoSumSmaller(nums, i + 1, target - nums[i]);
        }
        return total;
    }
    
    int twoSumSmaller(vector<int>& nums, int start, int target) {
        int left = start, right = nums.size() - 1, count = 0;
        while (left < right) {
            if (nums[left] + nums[right] < target) {
                count += (right - left);
                left++;
            } else {
                right--;
            }
        }
        return count;
    }
};